app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/chart/BeaconLineChart.java

```diff
@@ -317,75 +321,80 @@ protected void drawBeacons(Canvas canvas) {
+                );
+            }
+        } else {
+            lastLinePoint = new PointF();
         }
 
+        lastLinePoint.x = currentLinePoint.x;
+        lastLinePoint.y = currentLinePoint.y;
+        lastAdvertisingPacket = advertisingPacket;
+    }
+
+    protected void fadeLastPoint(Canvas canvas) {
         if (lastLinePoint != null) {
             linePaint.setAlpha(255);
             canvas.drawCircle(lastLinePoint.x, lastLinePoint.y, pixelsPerDip * 4, linePaint);
         }
     }
 
-    protected float getValue(Beacon beacon, AdvertisingPacket advertisingPacket, long windowLength) {
-        List<AdvertisingPacket> recentAdvertisingPackets = null;
-        int[] recentRssis;
-        float meanRssi;
-
-        // make sure that the window size is at least 10 seconds when we're looking for the frequency
-        windowLength = (valueType != VALUE_TYPE_FREQUENCY) ? windowLength : Math.max(windowLength, 10000);
-
-        if (windowLength == 0) {
-            meanRssi = advertisingPacket.getRssi();
-        } else {
-            recentAdvertisingPackets = beacon.getAdvertisingPacketsBetween(
-                    advertisingPacket.getTimestamp() - windowLength,
-                    advertisingPacket.getTimestamp()
-            );
-            recentRssis = AdvertisingPacketUtil.getRssisFromAdvertisingPackets(recentAdvertisingPackets);
-            meanRssi = AdvertisingPacketUtil.getMeanRssi(recentRssis);
-        }
+    protected float getValue(Beacon beacon, AdvertisingPacket advertisingPacket) {
+        int rssi = advertisingPacket.getRssi();
+        return processReturnValue(beacon,advertisingPacket, rssi);
+    }
 
+    protected float processReturnValue(Beacon beacon, AdvertisingPacket advertisingPacket, float rssi) {
         switch (valueType) {
             case VALUE_TYPE_RSSI: {
-                return meanRssi;
+                return rssi;
             }
             case VALUE_TYPE_DISTANCE: {
-                return BeaconDistanceCalculator.calculateDistanceTo(beacon, meanRssi);
+                return BeaconDistanceCalculator.calculateDistanceTo(beacon, rssi);
             }
             case VALUE_TYPE_FREQUENCY: {
-                return 1000 * (recentAdvertisingPackets.size() / (float) windowLength);
+                // make sure that the window size is at least 10 seconds when we're looking for the frequency
+                long windowLength = (valueType != VALUE_TYPE_FREQUENCY) ? this.windowLength : Math.max(this.windowLength, MINIMUM_WINDOW_LENGTH_FREQUENCY);
+                List<AdvertisingPacket> recentAdvertisingPackets = beacon.getAdvertisingPacketsBetween(
+                        advertisingPacket.getTimestamp() - windowLength,
+                        advertisingPacket.getTimestamp()
+                );
+                // convert frequency from milliseconds to seconds to receive Hertz
+                return TimeUnit.SECONDS.toMillis(1) * (recentAdvertisingPackets.size() / (float) windowLength);
```

Comment:
Possible `ArithmeticException` for dividing by zero if `windowLength` is zero.

Reply:
windowLength is never zero